lighthouse-core/scripts/generate-lh-timing-trace.js

```diff
@@ -0,0 +1,129 @@
+/**
+ * @license Copyright 2016 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+
+/* eslint-disable no-console */
+
+/**
+ * Generates a chromium trace file from user timing measures
+ * Adapted from https://github.com/tdresser/performance-observer-tracing
+ * @param {!Array{{}}} entries user timing entries
+ */
+function generateTraceEvents(entries) {
+  const currentTrace = [];
+  let id = 0;
+
+  for (const entry of entries) {
+    const traceEvent = {
+      name: entry.name,
+      cat: entry.entryType,
+      ts: entry.startTime * 1000,
+      dur: entry.duration * 1000,
+    };
+
+    traceEvent.pid = 'Measurements';
+
+    switch (entry.entryType) {
+      case 'mark':
+        traceEvent.pid = 'Marks';
+        break;
+      case 'measure':
+        if (entry.name.startsWith('audit-')) traceEvent.tid = 'Audits';
+        else if (entry.name.startsWith('gather-')) traceEvent.tid = 'Gatherers';
+        else traceEvent.tid = 'TopLevelMeasures';
+        break;
+      default:
+        traceEvent.pid = 'Primary';
+    }
+
+    if (entry.entryType === 'resource') {
+      entry.url = traceEvent.name;
+      traceEvent.name = 'resource';
+    }
+
+    if (entry.duration === 0) {
+      traceEvent.ph = 'n';
+      traceEvent.s = 't';
+    } else {
+      traceEvent.ph = 'X';
+    }
+
+    traceEvent.id = '0x' + id.toString(16);
+    id++;
+
+    const args = {};
+    for (const key of Object.keys(entry)) {
+      const value = entry[key];
+      if (key === 'entryType' || key === 'name' || key === 'toJSON') {
+        continue;
+      }
+      args[key] = value;
+    }
+    traceEvent.args = args;
+
+    currentTrace.push(traceEvent);
+  }
+
+  return currentTrace;
+}
+
+/**
+ * Writes a trace file to disk
+ * @param {!Array{{}}} entries user timing entries
+ * @param {?string} traceFilePath where to save the trace file
+ */
+function saveTraceOfTimings(entries, traceFilePath) {
+  const events = generateTraceEvents(entries);
+
+  const jsonStr = `
+  { "traceEvents": [
+    ${events.map(evt => JSON.stringify(evt)).join(',\n')}
```

Comment:
haha how about `JSON.stringify({traceEvents: events}, null, 2)` :)

Reply:
well the benefit of stringifying them individually is 1 line per trace event: 

![image](https://user-images.githubusercontent.com/39191/32400203-70955a4a-c0ba-11e7-9d0d-2efcd694c25c.png)

sooooooo........

Reply:
alright alright